### DISCUSSION TITLE
isZero() does not respect config.epsilon

### DISCUSSION THREAD

**Discussion by gwhitney** [josdejong/mathjs#2838] [Design decisions]
This concern came up in the implementation of the polynomial root solver that I am doing as a basic algebraic manipulation benchmark for mathjs/picomath. I was using `isZero` to test the discriminant of quadratics/cubics, but it involves some cancellation, so it can come out to a minuscule technically non-zero floating point number. I had assumed that `isZero` was a "relational function" as described in the mathjs docs and hence would consider such a result to be zero. However, that's not currently the case, leading to erroneous results (two distinct roots when there really should be one, or complex roots when they should be real, for example). So should we:

(a) Modify isZero to respect config.epsilon? (Clearly a breaking change...)
(b) Add an equalZero function? (Right now one can use `equal(x, 0)` but that leads to an unnecessarily complicated double dispatch.)

**Comment by gwhitney** [comment]
Oh, actually all that `equal(x, 0)` tests is whether x is less than Number.EPSILON (the smallest number you can add to one that changes it). Not sure why that makes sense, but right now it has absolutely nothing to do with config.epsilon. Since this is a **relative** comparison, perhaps no x should be `equal(x, 0)` other than 0 itself, because the difference is always equal to x, and hence not small relative to x. I also don't understand the logic: if two numbers are within Number.EPSILON of each other, they are considered equal. But 1e-19 and 1e-20 are quite distinct (the difference is 90% of the larger) and floating point numbers have no trouble distinguishing them, so considering them to be relatively equal doesn't make any sense. There is a line in the documentation about how equal() can't be used to compare numbers smaller than Number.EPSILON, and it is the case, but only because of the convention that two numbers within Number.EPSILON are always equal regardless of config.epsilon, violating the "relative" aspect of the comparison. I think this violation comes from a desire for it to be the case that if `equal(a,b)`, then `equal(a-b, 0)` -- but that's not even true right now, and it doesn't actually make sense for relative comparison.

So my thinking at the moment: This violation is unnecessary and the situation is buggy. Change equal to truly be relative, so that the only thing equal to 0 is actual 0. Perhaps add an isNegligible function to mean "is less than config.epsilon in absolute value" -- that may need to be used in some places that relying on the "within Number.EPSILON is equal" fallback.

But I didn't want to file a bug until I could get at least a second opinion. Thoughts?

**Reply by josdejong** [reply]
Good point. It would be logic indeed if `isZero` would be consistent with `equal` and utilizes `config.epsilon`. I'm open to discuss ways to improve this!

For reference: the current implementation of `nearlyEqual` used by the relational functions is a mix of an absolute comparison and a relative comparison: 
https://github.com/josdejong/mathjs/blob/77f94c4f95f9238593f8dd0bd3bf31b27248e7a9/src/utils/number.js#L598-L636

Unfortunately I didn't put any reference on the formulas and where I found them, I remember I did some reading up on comparing numbers and that this can be challenging and some background reading can help understanding the problem. Calculating the relative difference works very well for any number except zero, since the comparison can only do an absolute comparison there, not a relative one.

So I think we do have two open ideas here:

1. Adjust `isZero` to utilize `epsilon` and become consistent with `equal`. I would say yes lets do that.
2. Can we adjust `equal` to work better with 0, and make `equal(a-b, 0)` work. I'm not sure about how to make this work though. In #2752 I mention a workaround , but I think this has problems of it's own.

See also: #2752, #1614

**Reply by gwhitney** [reply]
> the current implementation of nearlyEqual used by the relational functions is a mix of an absolute comparison and a relative comparison

Gotcha. The documentation is not clear on this, it just says it is relative comparison, so at least documentation and reality ought to be synced.

> Adjust isZero to utilize epsilon and become consistent with equal. I would say yes lets do that.

Well, but `equal(x, 0)` makes no use of config.epsilon at the moment. It uses Number.EPSILON (and is not at the moment configurable). We could change isZero to use Number.EPSILON as well to make them consistent, but it is not clear why Number.EPSILON is relevant to comparing to zero, since it has to do with numbers near one, not near zero.

> make equal(a-b, 0) work
If someone does
```
let x = a - b
if (math.equal(x, 0)) ...
```
and expects that to work like
```
if (math.equal(a, b)) ...
```
then I think they are simply out of luck in a world where we are doing relative comparisons, since all scale information is lost by the time you get to `equal(x,0)`.

In light of all of the above, here are some measures we could take that might help:

Either or both of:
1. Add an isNeglible function that just tells you if something is less in absolute value than config.epsilon.
2. Change nearlyEqual to be purely relative, no absolute component, no involvement of Number.EPSILON. In this case, `equal(x, 0)` will come to match `isZero(x)` and be true only when x is precisely zero. (That behavior for zero may admittedly be unintuitive for casual users.)

Alternatively:
- Overhaul the approximate comparison system for one that is deliberately and configurably a mix of absolute and relative comparison, with separate relative_epsilon and absolute_epsilon settings. Two numbers would be considered equal if they are equal either because of the relative_epsilon or absolute_epsilon. Setting relative_epsilon to 0 would give pure absolute-difference comparison, as the poster in #2752 seemed to want, and conversely setting absolute_epsilon to 0 would give pure relative-difference comparison, as you likely want if you are dealing with small quantities. Defaulting relative_epsilon to the current default for config.epsilon and absolute_epsilon to Number.EPSILON would more or less exactly recover the current behavior, which does seem to generally work basically OK in a rough-and-ready way for everyday purposes.

In either case perhaps there should be a way to do some comparisons with different epsilon values than the current config? Or is it easy and not computationally expensive to just set config values to something else during a subcomputation and then set them back?

Either route should be combined with carefully documenting the situation.

I guess at the moment I'd lean toward the latter alternative because it involves no new functions, seems to give everyone what they want, and is not a breaking change (except for changing the name config.epsilon to config.relative_epsilon; we could temporarily make config.epsilon an alias for config.relative_epsilon that issues a deprecation warning, and at a future breaking change make it an error to access config.epsilon).

**Comment by josdejong** [comment]
We definitely have to at least improve the documentation about it 😅 .

> Well, but `equal(x, 0)` makes no use of config.epsilon at the moment

O wow, now I see. I was at first assuming it was also using `config.epsilon` but it is indeed using `Number.EPSILON`. Anyway, when removing this absolute comparison with DBL_EPSILON, some of the following unit tests fail. Looking at the FIXME's these tests and logic can use some love anyway:

https://github.com/josdejong/mathjs/blob/b050d5ed5d2c9f35900a68d350e84bac1ccb4a82/test/unit-tests/utils/number.test.js#L546-L559

**1.**

Yes makes sense to me to introduce two variants of `isZero`, a "nearly equal" variant and an "exact" variant. I have the feeling that it would be most consistent to make `isZero` part of the relational functions, using `epsilon` just like all the other functions like `equal` and `smaller`. So, maybe we can do that, an if needed (depending on the decisions in (2) and (3)), create an `isStrictlyZero()` variant.

**2.**

The current behavior of comparing to zero is weird anyway, so I think it is a good idea to improve that. It makes sense to me to introduce a `relativeEpsilon` (the current `epsilon`) and an `absoluteEpsilon`. For backward compatibility we should set `absoluteEpsilon = Number.EPSILON`, but I think in general a better setting would be `absoluteEpsilon = 1e-12`. I think working with "regular" values and wanting to compare `equal(a-b,0)` is much more common in real world usage than working with extremely small values like `1.4e-21`, so maybe best to cater for that. What do you think?

**3.**

The function `isZero` is not the only function that we may want to change to use epsilon. We also have: `isPositive`, `isNegative`, and `isInteger`. I think those functions should ideally use epsilon too, and become a "real" part of the relational functions. Maybe we should also move them from `/utils` to `/function/relational`.

**Reply by gwhitney** [reply]
1. Agreed if the approximate comparisons are regularized and documented and made systematic with explicit hybrid relative/absolute behavior, and isZero is made a relational function equivalent to equal(x, 0), then there is no need to add another function (`x === 0` would serve most cases of isStrictlyZero anyway). But this is definitely a breaking change on isZero, and so could only be done on a major version bump.
2. Well, my vote would be for absoluteEpsilon to be at least a few orders of magnitude smaller than relativeEpsilon so that the relative behavior doesn't start to erode right at 1, but rather more like at .0001 or so, since people do more frequently use fractions/decimals down to 1/1000 or a bit smaller. So if relativeEpsilon is kept at 1e-12 then I'd say 1e-16 is just about right for absoluteEpsilon, and so if you wanted to make the shift initially non-breaking it would be fine to just (start with, anyway) using Number.EPSILON which is in that ballpark anyway. Then if isZero is changed in a subsequent breaking change, that could be an opportunity to simplify to 1e-12/1e-16 (or whatever is chosen) and remove the slightly dubious connection with Number.EPSILON.
3. Yes these other functions could with care be made approximate relational as well. I say with care because you never want -1e-17 being reported as positive (just because it is nearly so). Rather, since in this world envisioned isZero will be true on it, you want both isPositive and  isNegative false on it. And you have to make sure that any number for which isInteger is true gets treated exactly like the integer that it is nearlyEqual to in all functions that care about integers (which may actually be a noticeable bit of work!) These are of course breaking changes, and should be done (or not) in parallel with isZero.

**Reply by josdejong** [reply]
1. 👍 . Yeah so let's not implement an `isStrictlyZero` right now, but await if there really is need for it.
2. makes sense. Since a floating point number has up to _almost_ 16 digits, 1e-16 feels like on the edge for me, but we can use that as a starting point and see if it indeed works out. 
3. yes indeed, the functions should be consistent with each other and not "overlap" in weird ways, good point.

So, to summarize: 

a. We want to change the internal `nearlyEqual` to check whether two values are relatively _or_ absolutely equal. To do this, we use a `relativeEpsilon=1e-12` (the old `epsilon`) and a new `absoluteEpsilon=1e-16`.
b. We want to refactor `isZero` `isPositive`, `isNegative`, and `isInteger` such that they become real relational functions, using `nearlyEqual ` like all other relational functions.

These are breaking changes.

**Reply by gwhitney** [reply]
Well (a) is an extremely slightly breaking change: 0 and 2e-16 will no longer be nearlyEqual, whereas they were before.  (Only numbers with absolute values in the very small interval (1e-16, 2.77e-16)  change in this way.) If one wants to do (a) as not a breaking change (at least at first), absoluteEpsilon could have the value Number.EPSILON by default (at least at first).

**Reply by dvd101x** [reply]
I was searching for references of the current relative absolute formula and found the following.

- [THE FLOATING-POINT GUIDE:](https://floating-point-gui.de/errors/comparison/) An explanation of what are the issues of absolute and relative. A proposal on how to combine them in an algorithm.
- [Python math.isclose:](https://docs.python.org/3/library/math.html#math.isclose): Apparently they had some [discussions](https://peps.python.org/pep-0485/#proposed-implementation) and implemented the following combination`abs(a-b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)`
- [Julia isapprox:](https://docs.julialang.org/en/v1/base/math/#Base.isapprox) Explained in the documentation and implemented this formula `norm(x-y) <= max(atol, rtol*max(norm(x), norm(y)))`

Python and Julia seems to have converged to a solution combining both.

**Reply by gwhitney** [reply]
Thanks for looking at the comparables! I have no objection to using the python-julia formula in mathjs.

**Reply by josdejong** [reply]
Thanks for your research David. I'm also openn to change the `nearlyEqual` formula used in mathjs for something better. So what would effectively change if we change to the approach used by Python and Julia?

**Reply by dvd101x** [reply]
I did a quick run of changing `nearlyEqual` and it looks promising. The main difference is that instead of taking `epsilon` as an argument, it would need `atol` and `rtol` it could be with the values previously discussed. It runs but breakers some tests.

Maybe a function should be created as it makes sense also for matrices or units as input.

As a side note `solveODE` could benefit of that function as currently it only takes `atol`

**Reply by josdejong** [reply]
Sounds good!

**Reply by gwhitney** [reply]
I definitely think that absoluteTolerance and relativeTolerance  (or whatever we want the "official" names to be) should become configuration parameters, with default values that minimize the number of breaking unit tests. The ones that do break will need to be examined by hand and decide whether to ship changes before or along with any PR that implements this change. If I understand correctly, epsilon then becomes an unused parameter; or is it still used somewhere else? If not, we may still want to deprecate but allow changes to the epsilon configuration, which would result in roughly corresponding changes to absoluteTolerance and relativeTolerance, for one release cycle to allow clients of mathjs to adjust to the change.

**Reply by dvd101x** [reply]
Sounds good. I will gladly get the ball rolling over the weekend for all to review the failing tests.

**Reply by josdejong** [reply]
@glen I think `epsilon` should become deprecated in the end, though a deprecation period + backward compatibility would be nice (if possible).

**Reply by josdejong** [reply]
mathjs `v13.0.0` now has `absTol` and `relTol` and the config is used by `isZero` see #3152, #3139.
